### ISSUE TITLE
How to throw exception from java to javaScript using j2v8 in android

### ISSUE THREAD

**Issue by rahulgiradkar00** [eclipsesource/J2V8#125]
How to throw exception from java to javaScript using j2v8 in android.

``` java

protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        V8 v8 = V8.createV8Runtime();
        v8.registerJavaMethod(new Printer(), "print");
        v8.executeVoidScript("try { " + "  print(null);" + "} catch (e) {" + "  print(e);" + "}");
    }

public static class Printer implements JavaCallback {
        @Override
        public  Object invoke(V8Object arg0, V8Array arg12) {
            Object arg1 = V8ObjectUtils.getValue(arg12, 0);
            if (arg1 == null)
                throw new NullPointerException("Naughty Developer!");
            System.out.println(arg1);
            return arg1;
        }
}
```

I wrote some java callback whose invoke method returns the object.
I was tried to throw java exception in javascript but application was crash was crash with following error.
Error : VM aborting Fatal signal 6 (SIGABRT) at 0x00000414 (code=-6), thread 1044 (com.demo)

**Comment by irbull** [comment]
This should work. I've converted this to a unit tests and it works as expected:
```
    @Test
    public void testExample() {
        V8 v8 = V8.createV8Runtime();
        v8.registerJavaMethod(new Printer(), "print");
        v8.executeVoidScript("try { " + "  print(null);" + "} catch (e) {" + "  print(e);" + "}");
    }

    public static class Printer implements JavaCallback {
        @Override
        public Object invoke(final V8Object arg0, final V8Array arg12) {
            Object arg1 = V8ObjectUtils.getValue(arg12, 0);
            if (arg1 == null) {
                throw new NullPointerException("Naughty Developer!");
            }
            System.out.println(arg1);
            return arg1;
        }
    }
```
When I run this, I get:
```
Naughty Developer!
```
Do you know if this happens on a particular device? 

Also, can you post the entire dump / trace that you get.

**Comment by irbull** [comment]
Also, what version of J2V8 are you using?

**Comment by rahulgiradkar** [comment]
I am using emulator as well as <b>samsung note 4 device</b> and version of J2V8 is V 3.0.0 . But it gives the same error.
<error>
Error : VM aborting Fatal signal 6 (SIGABRT) at 0x00000414 (code=-6), thread 1044 (com.demo)</error>

Nothing is displayed on logcat.

**Comment by irbull** [comment]
Can you try with 3.1 (3.1.6 to be exact). There was a bug related to accessing freed memory in callbacks, and this looks similar. The packaging has changed a bit with 3.1. I wrote about how to consume it here [http://eclipsesource.com/blogs/2015/11/04/shipping-j2v8-as-an-aar/](http://eclipsesource.com/blogs/2015/11/04/shipping-j2v8-as-an-aar/)

**Comment by rahulgiradkar** [comment]
Still same error I was changed the J2V8 lib to 3.1 and tried on device too.

<code>
 E/dalvikvm(1847): VM aborting
12-29 01:15:54.870: A/libc(1847): Fatal signal 6 (SIGABRT) at 0x00000737 (code=-6), thread 1847 (.demoexample)
</code>


Case 1: The exception handling working properly when function doesnot return anything. 
i.e void class implement from JavaVoidCallback
```java
JavaVoidCallback voidCallback = new JavaVoidCallback() {
			@Override
			public void invoke(V8Object arg0, V8Array arg1) {
				Object value = V8ObjectUtils.getValue(arg1, 0);
				if (value == null)
					throw new NullPointerException("Null Value in JavaVoidCallback");
                               System.out.println(value);
			}
```


Case 2: The exception handling failed when function return any Object.
i.e. class implement from JavaCallback
```java
JavaCallback Callback = new JavaCallback() {
			@Override
			public Object invoke(V8Object arg0, V8Array arg1) {
				Object value = V8ObjectUtils.getValue(arg1, 0);
				if (value == null)
					throw new NullPointerException("Null Value in JavaVoidCallback");
                                System.out.println(value);
                                return value;
			}
```
###### Case 2 always failed. I was tried on different emulators as well as devices.

**Comment by irbull** [comment]
@rahulgiradkar Did you mean to delete your most recent comment? Java callbacks and JavaVoid callbacks handle the stack differently, so a bug here doesn't surprise me. I got a note about a new comment, but now it's gone.

**Comment by skmll** [comment]
Hi,

I am using 3.1.6 J2V8 and i also can not seem to make this work. Questions:

Can i throw a RuntimeException in a method registered reflectively?
(From what i could debug, th exception is caught in the "callObjectJavaMethod" under the InvocationTargetException clause)

If not ,can i throw a RuntimeException in a JavaCallback registered in a V8Object instead of the runtime?

Even registering the JavaCallback directly in the Runtime, it doesnt work for me. 
It keeps "VM aborting"

Thanks in advance

**Comment by irbull** [comment]
I think this was actually caused by checking for return values after exceptions were thrown. On some VMs those return values were garbage. This was fixed in 27a8d37bc08b6a09b10cf34e5226e4f167d34df8
